internals/slo.py

```diff
@@ -63,22 +63,68 @@ def remaining_days(requested_on: datetime.datetime, slo_limit: int) -> int:
   return slo_limit - weekdays_between(requested_on, now_utc())
 
 
-def record_vote(gate: Gate, votes: list[Vote]) -> bool:
+def record_vote(gate: Gate, votes: list[Vote], old_gate_state: int) -> bool:
   """Record a Gate SLO response time if needed.  Return True if changed."""
-  if gate.requested_on is None:
-    return False  # Review has not been requested yet.
-  elif gate.responded_on is not None:
-    return False  # We already recorded the time of the initial response.
-  elif not votes:
-    return False  # No votes yet
-  else:
-    recent_vote_time = max(v.set_on for v in votes)
-    if recent_vote_time > gate.requested_on:
-      logging.info('SLO: Got reviewer vote as initial response')
-      gate.responded_on = recent_vote_time
-      return True
+  if not votes:
+    return False
+  latest_vote = sorted(votes, key=lambda v: v.set_on)[-1]
+  latest_state = latest_vote.state
 
-  return False
+  if latest_state == Vote.NO_RESPONSE:
+    return False  # NO_RESPONSE never changes SLO state.
+
+  changed = False
+  if latest_state in Vote.REQUESTING_STATES:
+    if gate.requested_on is None:
+      logging.info('SLO: Someone requested a new review')
+      gate.requested_on = latest_vote.set_on
+      changed = True
+
+  if latest_state in Vote.RESPONSE_STATES:
+    if gate.responded_on is None:
+      logging.info('SLO: Got reviewer vote as initial response')
+      gate.responded_on = latest_vote.set_on
+      if gate.requested_on is None:
+        logging.info('SLO: Reviewer is the person initiating the review')
+        gate.requested_on = latest_vote.set_on
+      changed = True
+
+  sent_back_for_rework = (
```

Comment:
nit: are sent_back_for_rework, sent_back_for_rework and sent_back_for_rework mutually exclusive? If so, maybe we should define right before they are used and return early for readability:
```
sent_back_for_rework = ...
if sent_back_for_rework:
  return true
finished_rework = ...
```